x/mongo/driver/topology/pool.go

```diff
@@ -855,6 +856,7 @@ func (p *pool) createConnections(ctx context.Context, wg *sync.WaitGroup) {
 
 			_ = p.removeConnection(conn, event.ReasonError)
 			_ = p.closeConnection(conn)
+			w.readyMu.Unlock()
```

Comment:
If the goal is to remove the connection from the pool before returning an error to `checkOut`, can we move the call `p.removeConnection` to above `w.tryDeliver` instead of adding a mutex and a goroutine?

Reply:
Doing this does lower the likelihood of encountering the race condition, but it doesn't eliminate it. If there are any handshake errors, then "clear" will be called from the "handshakeErrFn" [here](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/pool.go#L853). The "handshakeErrFn" will call "tryDeliver" when emptying the idle connections wait queue [here](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/topology/pool.go#L686). Since this happens before the connection is closed, it will (probably) signal to the checkout function that the connection is "ready".

Reply:
My understanding is that the goal of this change is to make sure that a "Connection closed" log is always sent _before_ the corresponding "Connection checkout failed" log. Is that correct?

If so, I'm not I understand how calling `p.removeConnection` before `w.tryDeliver` could lead to any log race condition. Consider the following sequence of calls:
```go
// Remove the connection from the pool, which emits a
// "Connection closed" event/log.
_ = p.removeConnection(conn, event.ReasonError)

// Deliver an error to the waiting checkOut, which will eventually emit a
// "Connection checkout failed" event/log in another goroutine.
w.tryDeliver(nil, err)

// Call the handshake error handler, which may clear the pool.
if p.handshakeErrFn != nil {
	p.handshakeErrFn(err, conn.generation, conn.desc.ServiceID)
}

// Actually close the underlying socket.
_ = p.closeConnection(conn)
```
By the time `p.removeConnection` returns, it should already have sent the "Connection closed" log message. There should be no way for the "Connection checkout failed" log message to happen before a "Connection closed" log message.

Reply:
From our conversation offline, re-arranging the "removeConnection" before "tryDeliver" does fix the ordering for "connection closed" and "checkout failed". However, "handshakeErrFn" is responsible for clearing the pool, so this change will cause a different race condition between "connection closed" and "pool cleared" messages. The bottleneck in all of this is "tryDeliver", which is why it was chosen as the place to lock the "ready" channel close.